### PR TITLE
usbus/msc: add initial Mass Storage Class support

### PR DESCRIPTION
**Pull request by dylad**
### Contribution description

This PR adds the initial support for Mass Storage Class in USBUS. This PR relies on the RIOT MTD implementation to implement the Mass Storage Class support. With the provided test application, a MTD device will be accessible as a normal storage device on your host computer.
Read and Write operations are allowed.
Multiple LUNs are supported so several MTD devices can be exported through USB.
The MSC relies on SCSI protocol to operate.

Currently there are some limitations:
    Supported host : Linux & Windows (macOS is untested)
    MSC cannot be used if MTD page size > 4096
    MTD device must have at least 512 bytes of memory to be exported.

Please be aware that performance are not so great.

### Testing procedure
Flash `tests/usbus_msc` application on a board with at least one MTD device.
Once the shell has started, prepare one or several MTD devices to be exported using `add_lun` command.
Once ready, start the USB connection with `usb_attach`

All MTD exported should appear as` /dev/sdX` on Linux.

### Issues/PRs references
Supersede #15941

### CODE REVIEW

File: sys/include/usb/usbus/msc/scsi.h [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    uint8_t block_desc_len;     /**< Length of block descriptor */
+} msc_mode_parameter_pkt_t;
+
+/**
+ * @brief CBW Packet structure for (@ref SCSI_READ10) and
+ * (@ref SCSI_WRITE10) requests
+ *
+ * @see Read10 or Write10 Command from SCSI Primary Command
+ *
+ * @warning uint16_t and uint32_t will be BIG ENDIAN
```
Comment (benpicco) [inline sys/include/usb/usbus/msc/scsi.h:153]: better use `network_uint16_t`/`network_uint32_t` then

File: sys/usb/usbus/msc/msc.c [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+    msc->block_offset += len;
+    /* Decrement whole len */
+    msc->cmd.len -= len;
+
+    /* buffer is full, write it and point to new block if any */
+    if (msc->block_offset >= block_size) {
+        if (mtd->page_size <= block_size) {
+            /* Write one or multiple pages */
+            printf("WRITE:%d bytes\n", block_size);
+            ret = mtd_write_page(mtd, msc->buffer, msc->block, 0,
```
Comment (benpicco) [inline sys/usb/usbus/msc/msc.c:97]: Since you are writing a full page, you can use `mtd_write_page_raw()` and avoid the read-modify-write cycle

File: sys/usb/usbus/msc/msc.c [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+
+static void _read_xfer(usbus_msc_device_t *msc)
+{
+    uint8_t lun = msc->cmd.lun;
+    mtd_dev_t *mtd = msc->lun_dev[lun].mtd;
+    uint16_t block_size = msc->lun_dev[lun].block_size;
+    int ret;
+
+    /* Check if we have a block to read and transfer */
+    //printf("block_nb:%d\n", msc->block_nb);
```
Comment (benpicco) [inline sys/usb/usbus/msc/msc.c:145]: either `DEBUG()` or remove it

File: sys/usb/usbus/msc/msc.c [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+#include <errno.h>
+
+#define ENABLE_DEBUG    1
+#include "debug.h"
+
+#include "mtd.h"
+
+/* Internal buffer to handle size difference between MTD layer and USB
+   endpoint size as some MTD implementation (like mtd_sdcard) doesn't allow
+   endpoint size transfer */
```
Comment (benpicco) [inline sys/usb/usbus/msc/msc.c:36]: Huh the MTD layer should take care of splitting up the transfers.
What do you mean here?

File: sys/usb/usbus/msc/msc.c (hunk 2 above)
Comment (benpicco) [reply sys/usb/usbus/msc/msc.c:97]: Ah I see you actually want this for cases where the block size is > 4096.
We should probably add a fast-path to `mtd_write_page()` then to skip the read-modify-write cycle if the whole sector is written - #19252

**Review by benpicco — COMMENTED** [review]
I don't know much about the USB part, but I can take a look at the MTD part.

File: sys/usb/usbus/msc/msc.c (hunk 4 above)
Comment (dylad) [reply sys/usb/usbus/msc/msc.c:36]: This buffer is used to store data coming from the endpoint. Since USB endpoints transfers (for Full-Speed device) are capped at 64 bytes, this buffer stores each transfer until it reaches the MTD device flash page size.
For example, if the MTD page size is 512 bytes. We need to have 8 endpoints transfer of 64 bytes to fill this buffer so we can trigger a write to the MTD device.
Moreover, since I cannot report the real page size of the MTD device if it's less than 512 bytes (e.g for EEPROM). This buffer will then have to store multiples pages at once.
But I agree that the wording is weird. I can come up with something better.

File: sys/include/usb/usbus/msc/scsi.h (hunk 1 above)
Comment (dylad) [reply sys/include/usb/usbus/msc/scsi.h:153]: Good idea !
But the `network_*` label puzzle me a bit.
How about `be_uint16_t` and so instead ?

File: sys/usb/usbus/msc/msc.c (hunk 2 above)
Comment (dylad) [reply sys/usb/usbus/msc/msc.c:97]: I wanted to implement both `mtd_write_page()` and `mtd_write_page_raw()` for some time but since this PR stalled I think we should just stick to one of them for now.
But I don't mind to switch to `mtd_write_page_raw()` if you think it is more suitable.

File: sys/include/usb/usbus/msc/scsi.h (hunk 1 above)
Comment (benpicco) [reply sys/include/usb/usbus/msc/scsi.h:155]: > How about `be_uint16_t` and so instead?

uh I didn't know we had both - yea `network_uint16_t` is just a typedef of `be_uint16_t` anyway, I just had never seen the latter.

File: sys/usb/usbus/msc/msc.c (hunk 4 above)
Comment (benpicco) [reply sys/usb/usbus/msc/msc.c:36]: You don't *need* to do this if you erase the sector on the first block and always write a whole sector. (`mtd_write_page_raw()` will not to the read-modify-write but just write and you don't need to write the whole sector at once).

There might be a slight performance benefit to it, but I'm not sure.

If you are the sole user of the MTD device you could also use `mtd->work_area` which is a per-device, dynamically allocated sector-sized buffer.

That would be a bit hacky though - but I also don't see a problem with just writing 64 bytes at a time.

File: sys/usb/usbus/msc/scsi.c [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+{
+    usbus_msc_device_t *msc = container_of(handler, usbus_msc_device_t,
+                                           handler_ctrl);
+    msc_cbw_rw10_pkt_t *pkt = (msc_cbw_rw10_pkt_t *)cbw->cb;
+    mtd_dev_t *mtd = msc->lun_dev[lun].mtd;
+
+    /* Convert big endian data from SCSI to little endian */
+    /* Get first block number to read from */
+    if (mtd->page_size < 512) {
+        msc->block = byteorder_swapl(pkt->blk_addr.u32) * (512 / mtd->page_size);
```
Comment (benpicco) [inline sys/usb/usbus/msc/scsi.c:66]:
```suggestion
        msc->block = byteorder_ntohl(pkt->blk_addr) * (512 / mtd->page_size);
```

File: sys/usb/usbus/msc/scsi.c [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    }
+    else if (mtd->page_size > 4096) {
+        /* TODO: add support */
+        assert(0);
+    }
+    else {
+        msc->block = byteorder_swapl(pkt->blk_addr.u32);
```
Comment (benpicco) [inline sys/usb/usbus/msc/scsi.c:73]:
```suggestion
        msc->block = byteorder_ntohl(pkt->blk_addr);
```

File: sys/usb/usbus/msc/scsi.c [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+    uint16_t block_size = msc->lun_dev[lun].block_size;
+
+    /* Convert big endian data from SCSI to little endian */
+    /* Get first block number to read from */
+    if (mtd->page_size < 512) {
+        /* If MTD pagesize is less than 512 (e.g for EEPROM),
+         * apply some logic and let host believe this is a 512
+         * page size device and let RIOT write several page in
+         * a row to reach 512 bytes so everyone is happy */
+        msc->block = byteorder_swapl(pkt->blk_addr.u32) * (512 / mtd->page_size);
```
Comment (benpicco) [inline sys/usb/usbus/msc/scsi.c:97]:
```suggestion
        msc->block = byteorder_ntohl(pkt->blk_addr) * (512 / mtd->page_size);
```

File: sys/include/usb/usbus/msc/scsi.h (hunk 1 above)
Comment (dylad) [reply sys/include/usb/usbus/msc/scsi.h:155]: I've switched to `be_uint*_t` where applicable. I'll run some tests later this week to ensure I didn't break something.

File: sys/usb/usbus/msc/msc.c (hunk 4 above)
Comment (dylad) [reply sys/usb/usbus/msc/msc.c:36]: > You don't need to do this if you erase the sector on the first block and always write a whole sector. 

Unfortunately, I am not sure I can guarantee I'll always write a whole sector. USB Host will use the reported page size and doesn't know about the sector size of the MTD device. So it may write only one or a few pages.

Comment (benpicco) [reply sys/usb/usbus/msc/msc.c:36]: Then how do you decide when to flush?
Or are those writes just lost if the sector is not completed?

File: sys/usb/usbus/msc/scsi.c [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+    size_t len = sizeof(msc_read_fmt_capa_pkt_t);
+
+    /*TODO: useless here but need a per MTD block_size array...*/
+
+    /* Clear the exact amount of bytes needed for this transfer */
+    memset(pkt, 0, len);
+
+    /* Only support one list, size of the whole struct minus the 4 first bytes */
+    pkt->list_length = len - 4;
+    pkt->blk_nb = byteorder_ltobl((le_uint32_t)blk_nb);
```
Comment (benpicco) [inline sys/usb/usbus/msc/scsi.c:165]:
```suggestion
    pkt->blk_nb = byteorder_htons(blk_nb);
```

Comment (dylad) [reply sys/usb/usbus/msc/scsi.c:165]: It hurts the brain.

File: sys/usb/usbus/msc/msc.c (hunk 4 above)
Comment (dylad) [reply sys/usb/usbus/msc/msc.c:36]: > Or are those writes just lost if the sector is not completed?

No, USB Host expects the USB device to write the data that were send. So if the Host only write a page, we must do a read-modify-write operation on our side.

Comment (benpicco) [reply sys/usb/usbus/msc/msc.c:36]: Ah now I get it!
USB will always transfer a page¹ (in 64 byte chunks) but since we can only erase a sector, we have to do a read-modify write cycle.

Buffering the page brings that down to one.

[1] a 'USB page' that will usually be just as big or bigger then the sector size. In that case we don't need the RMW. Only if the MTD page is greater than 4096 bytes we have chose a smaller USB page and resort to RMW.

I would propose the following:

 - if the 'USB page' is smaller than the sector size, allocate a temporary buffer as you do here, but you may use `malloc` (as that's already used by `mtd_pagewise` - I think `malloc` on init is fine.
 - if the 'USB page' is greater or equal the sector size, don't allocate a buffer and write the bytes as you get them with `mtd_write_page_raw()`, ereasing the sector at the first block.

That way you don't have the memory overhead in the common case and don't have to tune config variables to fit the storage.

What do you think?

Comment (dylad) [reply sys/usb/usbus/msc/msc.c:36]: > USB will always transfer a page¹ (in 64 byte chunks) but since we can only erase a sector, we have to do a read-modify write cycle.

Very true.

> Buffering the page brings that down to one RMW cycle.

That's the idea.

To sum up, there are 3 cases to properly handle:

- USB reported page size == MTD page size

  In this case, (512, 1024, 2048, 4096), this is pretty easy. We store the 64 bytes chunks into a buffer (allocated dynamically or not) Since we have a whole MTD page (not a MTD sector) we can write it directly to the memory. (And we rely on the MTD implementation to handle the erase part).
- USB reported page size > MTD page size

This is the case for a few MTD type like EEPROM. In this case, we report an USB page size of 512 (the minimum allowed). So if MTD page size is 256. We have to write two MTD pages for each USB page. 
This is also why we cannot export MTD device which have less than 512 bytes of memory. (Well we could cheat but this will just add more complexity for a small corner case and I believe this PR is already complex enough...)

- USB reported page size < MTD page size

This is a more tricky one, which is not yet support by this PR (even though there is some code laying around). For this case, we need to do some RMW cycle. But this will belong to a followup PR.

Comment (dylad) [reply sys/usb/usbus/msc/msc.c:36]: > I would propose the following:
    if the 'USB page' is smaller than the sector size, allocate a temporary buffer as you do here, but you may use malloc (as that's already used by mtd_pagewise - I think malloc on init is fine.
    if the 'USB page' is greater or equal the sector size, don't allocate a buffer and write the bytes as you get them with mtd_write_page_raw(), ereasing the sector at the first block.

> That way you don't have the memory overhead in the common case and don't have to tune config variables to fit the storage.

> What do you think?

Performance-wise, the second idea is great but I need to check if it won't make the code unreadable because of the corner cases.
For the first idea, I think this will the most common pattern. I can switch to dynamically allocated buffer since we let the user chose which MTD to export through the test application. malloc() can then be call only and only if the user request the MTD device to be exported. This may save some memory on boards which have a several MTD that one can export but only need one to be exported.

Comment (benpicco) [reply sys/usb/usbus/msc/msc.c:36]: But why tie this to page size? A page is trust the largest transfer unit, but the MTD will already take care of that, so on the application level it's rather arbitrary.

Sector size is the much more interesting value as it's the erase unit. If you report the sector size (and you will always get a full sector through MSC) you can just erase the sector at the start of the transfer and then write the 64 byte blocks as they come (without the RMW).

Comment (dylad) [reply sys/usb/usbus/msc/msc.c:36]: I see your point. This will increase performance for some MTD devices. So you suggest to erase the sector from USBUS then call `mtd_write_page_raw()` for each USB endpoints chunks received to avoid storing the content of a page ?
I can try to give it a go. I don't think such rework would be too painful.
Obviously, this won't work at all for sector size > 4096.

Comment (dylad) [reply sys/usb/usbus/msc/msc.c:36]: I am making progress here, I have a working version using sector size instead of page size.
I am now looking at the buffer side and I think we shouldn't drop it even if the USB block == MTD sector size:
This might impact performance on read operation. Currently, a single `mtd_read()` is perform and store in the buffer. Then we transfer the buffer through USB chunks. Dropping the buffer would force us to call `mtd_read()` for each USB transfer chunks and it seems sub-optimal to me.

Regarding the dynamical allocation of the buffer, I am not against. This can be done during the `USB_MSC_SETUP_REQ_GML` request. This way we can iterate through all exported MTD devices to USB. Determine which device has the greatest sector size and allocate this sector size to the buffer. No need for user configuration.

Note that even if multiple LUNs are used at the same time. Operation can only be done one LUN at a time. So having one buffer is completely fine for all LUNs (read or write operations).

**Comment by dylad** [comment]
Lastest commit changes the following:
- Use MTD sector size instead of MTD page size as USB block.
- Dynamically allocate the internal buffer depending on the MTD with the greatest sector size value.

File: sys/include/usb/usbus/msc.h [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+ * @brief MSC initialization function
+ *
+ * @param   usbus   USBUS thread to use
+ * @param   handler MSC device struct
+ */
+
+int usbus_msc_init(usbus_t *usbus, usbus_msc_device_t *handler);
+
+/**
+ * @brief Register a MTD device as a MSC LUN
```
Comment (benpicco) [inline sys/include/usb/usbus/msc.h:91]: Would be good to have LUN explained at least once

File: sys/usb/usbus/msc/msc.c [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+{
+    (void)usbus;
+    (void)state;
+
+    usbus_msc_device_t *msc = container_of(handler, usbus_msc_device_t,
+                                           handler_ctrl);
+
+    switch (setup->request) {
+    case USB_MSC_SETUP_REQ_GML:
+        uint8_t data = _get_lun(msc);
```
Comment (benpicco) [inline sys/usb/usbus/msc/msc.c:367]:
```suggestion
        data = _get_lun(msc);
```

File: sys/usb/usbus/msc/msc.c [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+                            usbus_control_request_state_t state,
+                            usb_setup_t *setup)
+{
+    (void)usbus;
+    (void)state;
+
+    usbus_msc_device_t *msc = container_of(handler, usbus_msc_device_t,
+                                           handler_ctrl);
+
+    switch (setup->request) {
```
Comment (benpicco) [inline sys/usb/usbus/msc/msc.c:365]:
```suggestion
    uint8_t data;
    switch (setup->request) {
```

File: sys/include/usb/usbus/msc.h [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+#endif /* USBUS_MSC_EXPORTED_DEV */
+
+typedef enum {
+    WAITING,
+    WAIT_FOR_TRANSFER,
+    DATA_TRANSFER_IN,
+    DATA_TRANSFER_OUT,
+    DATA_TRANSFER_LAST,
+    GEN_CSW
+} usbus_msc_state_t;
```
Comment (benpicco) [inline sys/include/usb/usbus/msc.h:45]: Eh, doxygen is right that those should be documented.
Someone (or you yourself) touching this code a few years down the line will be thankful to have an explanation what those states mean.

File: sys/include/usb/usbus/msc.h [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+#include "usb/usbus.h"
+#include "usb/usbus/msc/scsi.h"
+#include "mtd.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef USBUS_MSC_EXPORTED_DEV
+#define USBUS_MSC_EXPORTED_DEV MTD_NUMOF
```
Comment (benpicco) [inline sys/include/usb/usbus/msc.h:35]: nitpick, but that sounds like you would say which device to export here.
How about
```suggestion
#define USBUS_MSC_EXPORTED_NUMOF MTD_NUMOF
```

File: sys/usb/usbus/msc/msc.c [hunk 14]
```diff
@@ -0,0 +1,453 @@
...
+                return -ENOMEM;
+            }
+            msc->buffer_size = block_size;
+            msc->lun_dev[i].mtd = dev;
+            return 0;
+        }
+    }
+
+    /* No more slots available */
+    return -EAGAIN;
```
Comment (benpicco) [inline sys/usb/usbus/msc/msc.c:263]:
```suggestion
    return -ENFILE;
```
might be a better fit

**Review by benpicco — COMMENTED** [review]
Nice one, works like a charm on my `same54-xpro` :smiley: 
Read speeds are ~600kB/s for both SD card and SPI flash, I wonder if this could benefit from #17091 as a follow-up.

Some doc comments mostly:

File: sys/include/usb/usbus/msc.h (hunk 9 above)
Comment (dylad) [reply sys/include/usb/usbus/msc.h]: Done here.

File: sys/include/usb/usbus/msc.h (hunk 12 above)
Comment (dylad) [reply sys/include/usb/usbus/msc.h]: Added comments and remove `DATA_TRANSFER_LAST` which was unused. Probably some leftover from long ago.

File: sys/include/usb/usbus/msc.h (hunk 13 above)
Comment (dylad) [reply sys/include/usb/usbus/msc.h]: Updated.

**Comment by dylad** [comment]
I think this PR would benefit #17091. At least, it will lower a bit the code complexity of this PR.
But since #17091 doesn't have any use case in master, I guess it's better to have this one first so @bergzand could give it a quick rebase right after ;)

File: sys/usb/usbus/msc/msc.c [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+            /* Clear relevant information */
+            msc->lun_dev[i].mtd = NULL;
+            msc->lun_dev[i].block_size = 0;
+            msc->lun_dev[i].block_nb = 0;
+            return 0;
+        }
+    }
+
+    /* No more slots available */
+    free(msc->buffer);
```
Comment (benpicco) [inline sys/usb/usbus/msc/msc.c:285]:
```suggestion
    free(msc->buffer);
    msc->buffer = NULL;
```

File: sys/usb/usbus/msc/scsi.c [hunk 16]
```diff
@@ -0,0 +1,386 @@
...
+
+    /* Clear the exact amount of bytes needed for this transfer */
+    memset(pkt, 0, len);
+
+    /* prepare pkt response */
+    pkt->type = 0;
+    pkt->removable = 1;
+    pkt->version = 0x01;
+    pkt->length = len - 4;
+    pkt->flags[0] = 0x80;
```
Comment (benpicco) [inline sys/usb/usbus/msc/scsi.c:135]: What flag is `0x80`?

File: sys/usb/usbus/msc/scsi.c [hunk 17]
```diff
@@ -0,0 +1,386 @@
...
+    size_t len = sizeof(msc_inquiry_pkt_t);
+
+    /* Clear the exact amount of bytes needed for this transfer */
+    memset(pkt, 0, len);
+
+    /* prepare pkt response */
+    pkt->type = 0;
+    pkt->removable = 1;
+    pkt->version = 0x01;
+    pkt->length = len - 4;
```
Comment (benpicco) [inline sys/usb/usbus/msc/scsi.c:134]: Why `- 4`?

File: sys/usb/usbus/msc/scsi.c [hunk 18]
```diff
@@ -0,0 +1,386 @@
...
+
+    /* Clear the exact amount of bytes needed for this transfer */
+    memset(pkt, 0, len);
+
+    pkt->error_code = SCSI_REQUEST_SENSE_ERROR;
+    pkt->sense_key = SCSI_SENSE_KEY_ILLEGAL_REQUEST;
+    /* Report the size of the struct minus the 7 first bytes */
+    pkt->add_len = len - 7;
+    pkt->asc = 0x30;
+    pkt->ascq = 0x01;
```
Comment (benpicco) [inline sys/usb/usbus/msc/scsi.c:255]: What are those magic numbers?

**Review by benpicco — COMMENTED** [review]
Please squash (and rebase)

btw, https://github.com/RIOT-OS/RIOT/pull/17091 implements URBs for CDC ECM.

File: sys/usb/usbus/msc/scsi.c (hunk 18 above)
Comment (dylad) [reply sys/usb/usbus/msc/scsi.c:255]: Is undocumented black magic numbers, a valid answer here ?

Comment (benpicco) [reply sys/usb/usbus/msc/scsi.c:255]: Where did you get them from?

File: sys/usb/usbus/msc/scsi.c (hunk 17 above)
Comment (dylad) [reply sys/usb/usbus/msc/scsi.c:134]: This is the additional packet length minus the "default" data of this packet.
a SCSI inquiry response will always include the 4 first bytes, but you may have to pass additional information. Thus, you pass the number additional bytes through this field.

I'll add a comment.

File: sys/usb/usbus/msc/scsi.c (hunk 16 above)
Comment (dylad) [reply sys/usb/usbus/msc/scsi.c:135]: This flag is SCCS (SCC Supported) bit. IIRC, this means Storage Controller Component.
This bit is commonly set but I forgot to create a `define` for this one

File: sys/usb/usbus/msc/scsi.c (hunk 18 above)
Comment (dylad) [reply sys/usb/usbus/msc/scsi.c:255]: Hmm, I think it was on some source code provided by ST for a STM32F1 that I used back in 2019.
I did have a look at Zephyr a few months ago too, but it was the same. (Just hexadecimal value used to fill their buffer).
I'll have a look at it tonight. I have a bunch of USB MSC related document on my home laptop. I'll see if I can give you more insight then.

File: sys/usb/usbus/msc/scsi.c (hunk 16 above)
Comment (dylad) [reply sys/usb/usbus/msc/scsi.c]: Was not needed in the end, so It got removed.

**Comment by dylad** [comment]
Rebased and squashed.
@benpicco Regarding the ACS, ACSQ "black magic" stuff: I've just add a comment for now.
These values are in fact used to report to the host additional information regarding different kind of failures.
I'll make use of this in a later PR, let's stick to default values for now, as more work is required.

**Review by benpicco — APPROVED** [review]

File: tests/usbus_msc/Makefile [hunk 19]
```diff
@@ -0,0 +1,29 @@
+
```
Comment (benpicco) [inline tests/usbus_msc/Makefile:1]: This will also need a `Makefile.ci`
You can create it with

    make generate-Makefile.ci

**Comment by benpicco** [comment]
bors merge

**Comment by benpicco** [comment]
bors cancel
bors merge

**Comment by benpicco** [comment]
bors try

**Comment by dylad** [comment]
Do we need a bors stop drinking again ?

**Comment by benpicco** [comment]
I think bors already went to sleep

bors ping

**Comment by dylad** [comment]
@benpicco Thanks for the review & merge !

File: tests/usbus_msc/main.c [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+
+    assert(usbdev);
+
+    /* Initialize basic usbus struct, don't start the thread yet */
+    usbus_init(&usbus, usbdev);
+
+    /* Initialize Mass Storage Class */
+    usbus_msc_init(&usbus, &msc);
+    /* Create USBUS thread */
+    usbus_create(_stack, USBUS_STACKSIZE, USBUS_PRIO, USBUS_TNAME, &usbus);
```
Comment (benpicco) [inline tests/usbus_msc/main.c:172]: Why is this needed here if we already have `auto_init_usbus`?

File: tests/usbus_msc/main.c [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+    puts("Then use the attach command to connect");
+    puts("your device and start USB operation");
+
+    /* Get driver context */
+    usbdev_t *usbdev = usbdev_get_ctx(0);
+
+    assert(usbdev);
+
+    /* Initialize basic usbus struct, don't start the thread yet */
+    usbus_init(&usbus, usbdev);
```
Comment (benpicco) [inline tests/usbus_msc/main.c:167]: Same here

File: tests/usbus_msc/main.c [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+
+    /* Initialize Mass Storage Class */
+    usbus_msc_init(&usbus, &msc);
```
Comment (benpicco) [inline tests/usbus_msc/main.c:170]: Should this be moved to `auto_init_usb.c`?

**Review by benpicco — COMMENTED** [review]
I've been trying to combine this with `stdio_cdc_acm` but so far no luck.

Is there some inherent conflict or is the issue just that we initialize the USBus twice?

**Comment by dylad** [comment]
Because It was easier to setup things this way.
The idea was to merge this stuff and let the users decided which MTD devices they want to use at runtime. And yes, it doesn't get along with CDC for now because I still need a bit of time to find the right solution for this.
Nevertheless, I should have a describe it in the Documentation, my bad.

I don't like the idea of initialize and export all MTD devices unconditionally. And I don't like the idea of an always ON MSC implementation either. It'll just increase the overall USB transfer for nothing with "ping" transfer aka TEST_UNIT_READY.

To me, the MSC and the MTD devices selection should be a user runtime choice.
Obviously, there are some uses cases to export all available MTD devices (this is why I implemented the multiple LUNs support) and to have a MSC always ON (hobbyst or academic playground). But I don't think it should be the default behavior.
